src/rgw/rgw_log.cc

```diff
@@ -317,32 +320,132 @@ void rgw_format_ops_log_entry(struct rgw_log_entry& entry, Formatter *formatter)
   formatter->close_section();
 }
 
-void OpsLogSocket::formatter_to_bl(bufferlist& bl)
+OpsLogManifold::~OpsLogManifold()
 {
-  stringstream ss;
-  formatter->flush(ss);
-  const string& s = ss.str();
+    for (const auto &sink : sinks) {
+        delete sink;
+    }
+}
 
-  bl.append(s);
+void OpsLogManifold::add_sink(OpsLogSink *sink)
+{
+    sinks.push_back(sink);
 }
 
-void OpsLogSocket::init_connection(bufferlist& bl)
+int OpsLogManifold::log(struct req_state* s, struct rgw_log_entry& entry)
 {
-  bl.append("[");
+  int ret = 0;
+  for (const auto &sink : sinks) {
+    if (sink->log(s, entry) < 0) {
+      ret = -1;
+    }
+  }
+  return ret;
 }
 
-OpsLogSocket::OpsLogSocket(CephContext *cct, uint64_t _backlog) : OutputDataSocket(cct, _backlog)
+OpsLogFile::OpsLogFile(CephContext* cct, std::string& path, uint64_t max_data_size) :
+  cct(cct), file(path, std::ofstream::app), max_data_size(max_data_size), data_size(0)
 {
-  formatter = new JSONFormatter;
-  delim.append(",\n");
 }
 
-OpsLogSocket::~OpsLogSocket()
+void OpsLogFile::flush()
+{
+  std::scoped_lock flush_lock(flush_mutex);
```

Comment:
can the log buffer append while the flush buffer is flushed?

Reply:
I think iow that maybe you could get that behavior by introducing a flag and dropping this mutex during flush (apologies if I'm missing something)

Reply:
The log_mutex is only held while swapping log_buffer and flush_buffer, so request threads can continue to append to the log buffer while the flush_buffer is actually being written out to disk. There should be minimal contention here between the loggers and the flusher. Let me know if I'm missing something, though! :)

Reply:
ok, i see;  this is basically fine, I think.